v2/bundle/spiffebundle/bundle.go

```diff
@@ -3,175 +3,411 @@ package spiffebundle
+)
+
+type bundleDoc struct {
+	jose.JSONWebKeySet
+	SequenceNumber uint64 `json:"spiffe_sequence,omitempty"`
+	RefreshHint    int64  `json:"spiffe_refresh_hint,omitempty"`
+}
+
 // Bundle is a collection of trusted public key material for a trust domain,
-// confirming to the SPIFFE Bundle Format as part of the SPIFFE Trust Domain
+// conforming to the SPIFFE Bundle Format as part of the SPIFFE Trust Domain
 // and Bundle specification:
 // https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE_Trust_Domain_and_Bundle.md
 type Bundle struct {
+	trustDomain spiffeid.TrustDomain
+
+	mtx            sync.RWMutex
+	refreshHint    *int64
+	sequenceNumber *uint64
+	jwtKeys        map[string]crypto.PublicKey
+	x509Roots      []*x509.Certificate
 }
 
 // New creates a new bundle.
 func New(trustDomain spiffeid.TrustDomain) *Bundle {
-	panic("not implemented")
+	return &Bundle{
+		trustDomain: trustDomain,
+		jwtKeys:     make(map[string]crypto.PublicKey),
+	}
 }
 
 // Load loads a bundle from a file on disk.
 func Load(trustDomain spiffeid.TrustDomain, path string) (*Bundle, error) {
-	panic("not implemented")
+	bundleBytes, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, spiffebundleErr.New("unable to read SPIFFE bundle: %w", err)
+	}
+
+	return Parse(trustDomain, bundleBytes)
 }
 
 // Read decodes a bundle from a reader.
 func Read(trustDomain spiffeid.TrustDomain, r io.Reader) (*Bundle, error) {
-	panic("not implemented")
+	b, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, spiffebundleErr.New("unable to read: %v", err)
+	}
+
+	return Parse(trustDomain, b)
 }
 
 // Parse parses a bundle from bytes.
-func Parse(trustDomain spiffeid.TrustDomain, b []byte) (*Bundle, error) {
-	panic("not implemented")
+func Parse(trustDomain spiffeid.TrustDomain, bundleBytes []byte) (*Bundle, error) {
+	jwks := &bundleDoc{}
+	if err := json.Unmarshal(bundleBytes, jwks); err != nil {
+		return nil, spiffebundleErr.New("unable to parse JWKS: %v", err)
+	}
+
+	bundle := New(trustDomain)
+	if jwks.RefreshHint > 0 {
+		bundle.refreshHint = &jwks.RefreshHint
+	}
+	if jwks.SequenceNumber > 0 {
+		bundle.sequenceNumber = &jwks.SequenceNumber
+	}
+
+	if jwks.Keys == nil {
+		// The parameter keys MUST be present.
+		// https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE_Trust_Domain_and_Bundle.md#413-keys
+		return nil, spiffebundleErr.New("no keys found")
+	}
+	for i, key := range jwks.Keys {
+		switch key.Use {
+		// Two SVID types are supported: x509-svid and jwt-svid.
+		case x509SVIDUse:
+			if len(key.Certificates) != 1 {
+				return nil, spiffebundleErr.New("expected a single certificate in %s entry %d; got %d", x509SVIDUse, i, len(key.Certificates))
+			}
+			bundle.AddX509Root(key.Certificates[0])
+		case jwtSVIDUse:
+			if err := bundle.AddJWTKey(key.KeyID, key.Key); err != nil {
+				return nil, spiffebundleErr.New("error adding key %d of JWKS: %v", i, errs.Unwrap(err))
+			}
+		case "":
```

Comment:
I think this case can be removed. From the spec:
>  Clients encountering unknown use values MUST ignore the entire JWK element.

Reply:
It was not clear to me what was the right thing to do here, since the spec also says "The use parameter MUST be set". I'm ok removing this.

Reply:
Hmm. That's a good point. @evan2645, what do you think? Spec seems to say "fail if not set, and ignore if set to an unknown value"... ?

Reply:
I'm sorry that the spec has caused this confusion. As part of this exchange, suggestions on how to clarify the text (and/or a PR/issue in https://github.com/spiffe/spiffe) would be useful.

Yes, `use` MUST be set... but requiring it to be set should not by itself suggest a particular behavior if a validator finds that this is not the case. All SPIFFE bundle producers MUST set the `use` claim. SPIFFE bundle validators MUST ignore JWK entries with unknown use claims (including an empty or missing `use` claim).

Reply:
If we have logging available, this would be a good place to do it.

Reply:
That makes sense. I think it is reasonable to require the `use` field as a condition of conformance and ignoring unknown `use` values seems like a condition of forwards compatibility.

Reply:
>ignoring unknown use values seems like a condition of forwards compatibility

FWIW, I believe this is actually a requirement stipulated by the JWKS RFC. Trying to do something with an unknown key type could be dangerous

Reply:
Yes, I agree ignoring unknown use values is ok. I'm still unsure if, as a validator, we should ignore empty or unset use values. I'm considering the experience when somebody tries to load a standard JWKS as a SPIFFE bundle. If we ignore all of the keys, because use isn't set, the "Load" will still succeed, albeit with an empty bundle...

Reply:
>I'm considering the experience when somebody tries to load a standard JWKS as a SPIFFE bundle. If we ignore all of the keys, because use isn't set, the "Load" will still succeed, albeit with an empty bundle...

In this case, it's not a SPIFFE bundle, it's a JWT-SVID bundle and needs to be loaded as such. We should also be checking `use` there for conformance with JWKS RFC... assuming our JWKS library doesn't already do it for us.

If someone tries to Load a SPIFFE bundle that doesn't have any SPIFFE keys in it... return an error?

Reply:
> it's a JWT-SVID bundle and needs to be loaded as such
Yes, I was more thinking of diagnostics when somebody does this on accident.

> If someone tries to Load a SPIFFE bundle that doesn't have any SPIFFE keys in it... return an error?
Is that something allowed by the spec? JWKS RFC only says that the keys array must exist, not that it must have at least one key...
> A JWK Set is a JSON object that represents a set of JWKs.  The JSON
   object MUST have a "keys" member, with its value being an array of
   JWKs.

Reply:
> We should also be checking use there for conformance with JWKS RFC... assuming our JWKS library doesn't already do it for us.
The `jwtbundle.Bundle` type assumes that the use is `jwt-svid, it does not require any SPIFFE specific parameters when parsing a JWKS document and does not set the `use` when marshalling. That's the current behavior in this branch.

> If someone tries to Load a SPIFFE bundle that doesn't have any SPIFFE keys in it... return an error?
Yes, that's checked in this PR and an error is returned in that case.

Reply:
> Is that something allowed by the spec? JWKS RFC only says that the keys array must exist, not that it must have at least one key...

Fair point. It does seem like a pretty strange condition, though...

>The jwtbundle.Bundle type assumes that the use is jwt-svid, it does not require any SPIFFE specific parameters when parsing a JWKS document and does not set the use` when marshalling. That's the current behavior in this branch.

Yes I think that is the correct behavior. We should do whatever the JWKS RFC says when parsing jwt-svid bundle. IIRC the RFC says to ignore unknown values. One open question is, in this case, `jwt-svid` use value is non unknown, so should be be technically able to load a jwt-svid bundle given a spiffe bundle? Seems like a nicety. But I don't think we should not allow a JWKS bundle to be loaded as a SPIFFE bundle

Reply:
@evan2645 @azdagron From the feedback received, I would keep this check and require the `use` field as a condition of conformance to the SPIFFE Trust Domain and Bundle spec, while we ignore the entire JWK element in case of an unknown `use` value. Are we ok with that?

Reply:
@amartinezfayo. I think what we've settled with is to ignore keys with missing or unrecognized `use` values, i.e. don't enforce conformance as a validator, just do the safe thing (i.e. ignore keys that aren't safe to use).

We don't have a good story for library-level logging for this kind of thing.

Reply:
Thanks for the clarification @azdagron.

Reply:
@evan2645 I filed https://github.com/spiffe/spiffe/issues/121 to consider a clarification from the spec perspective.